src/coreclr/jit/importer.cpp

```diff
@@ -4801,6 +4802,10 @@ NamedIntrinsic Compiler::lookupNamedIntrinsic(CORINFO_METHOD_HANDLE method)
             {
                 result = NI_System_Math_Floor;
             }
+            else if (strcmp(methodName, "SinCos") == 0)
+            {
+                result = NI_System_Math_SinCos;
```

Comment:
Is the JIT doing anything interesting for this intrinsic? If not, I think we should do the hookup in the JIT only once it starts doing something interesting.

Also, I suspect that we will want to mark the public entrypoint that returns the tuple as the intrinsic since the JIT has problem representing operations that return multiple values as discussed in the `Math.DivRem` issue.

Reply:
> Is the JIT doing anything interesting for this intrinsic? 

No, as is the case with most of the Math APIs. We really only special-case `Sqrt`, `Abs`, `Round`, `Ceiling`, and `Floor` today.
We probably could/should handle all of them in CSE, but it would require we not enable it for crossgen scenarios due to existing cross-platform differences.

> Also, I suspect that we will want to mark the public entrypoint that returns the tuple as the intrinsic since the JIT has problem representing operations that return multiple values as discussed in the Math.DivRem issue.

I actually did this initially, but returning a `struct` broke on x86 and so I switched to the pointer approach instead since that's what the CRT implementation is.
If we decide to eventually normalize the math implementation across platforms, we might be able to revisit this, but until that happens (and until we can optimize the call into the CRT), I don't think we can move away from having a call to a method that takes pointers.

Reply:
> I actually did this initially, but returning a struct broke on x86 and so I switched to the pointer approach instead since that's what the CRT implementation is.

I agree that the FCall needs be returning the floats using pointer approach.

My comment was about which method is tagged with `[Intrinsic]` attribute and recognized by the JIT. I believe it will need to be the public method that returns the tuple for the JIT optimizations to work, not the FCall that returns the two values via pointers.

Reply:
x87 has an instruction to perform Sin & Cos I guess we don't want to emit that.

> We probably could/should handle all of them in CSE

Don't we already do it for all Math methods? E.g. [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYRoG8GuGbunGAZgBsI2DKwDKAUQAUw0eIQBKXt3aq+XYgHYGAWTEALAGIA6FhFwzlDAFT6jZi1Zv2DGE+cvWldhx6dvZQBuDQYAXzCw+gZ5MQYAGQBJFj05EXjlMPVqTW448QAvBgBeBgAGUNy82OgGGQBLADtxBpLKhgaAHnJyvuDOgGpBlWq8nJrNYsGyhr93T2cfKsnIsc0dBkKV7jXwoA=)
So we need to mark them as Intrinsic in order to assign a correct VN/mark as pure.

Reply:
You're right, we do handle CSE today via the named intrinsics. It was constant folding that I was thinking of which we don't do that and which we could/should (for a given run and not for AOT/Crossgen).
I also found a place I missed handling `NI_System_Math_SinCos` in gentree.cpp and updated it to handle that.

For handling in CSE however, I think I need some guidance from @briansull since this deals with a tuple return, which is a bit different from the existing handling in `valuenum`

Reply:
I don't think we should try to CSE this intrinsic.
Seem too unlikely to ever occur as a CSE in real world code.

Reply:
> Seem too unlikely to ever occur as a CSE in real world code.

I'm not so sure. Particularly when working with `Quaternions`, it is likely that you will compute `SinCos` across several calls.

So, all it takes is for someone to have `Matrix4x4.CreateRotationX(float)` and `Matrix4x4.CreateRotationY(float)` where `radians` is the same for both to result in a real world CSE case involving inlining.
Likewise, `SinCos(x)` could be used as a CSE for `Sin(x)` and Cos(x)`, which might also occur across inlines.

Reply:
I suspect that the CSE of these will require more substantial changes in the JIT. To make progress, it may be best to omit all JIT changes from this PR and leave them for follow up PR.

Reply:
Dropped the `NI_System_Math_SinCos` support since it requires more extensive changes. Will log a follow up issue to track the additional work to enable it.